### PR TITLE
update to AWS SDK PHP 3.x & add profile/region support

### PR DESCRIPTION
**Pull request by fdewinne**
This update adds the following features:
- authentication through profile instead of key/secret pair
- region support with auto-detection from environment or profile configuration if available
- fix bug if no key is provided

This update also provides unit tests:

> see https://continuousphp.com/git-hub/fdewinne/composer-aws/build/f4223295-2754-4ef4-96d1-21e956f3594a/output for execution details

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1 +1,3 @@
+.idea
 /vendor
+composer.lock
```
Comment (naderman) [inline .gitignore]: Please do not ignore the lock file, properly update it and commit it.

**Comment by fdewinne** [comment]
I just restored the composer.lock file... even if it's not necessary in this case.
If you look at the final note in the [composer.lock documentation](https://getcomposer.org/doc/01-basic-usage.md#composer-lock-the-lock-file), we are in the case of a library in a composer plugin. Then composer ignores this file on install command.

**Comment by naderman** [comment]
@fdewinne Yes but while the file is not needed for installing the library or plugin it has advantages to keep a lock file for libraries regardless: The use of composer install in CI is faster and deterministic (without the lock composer runs an `update` which it shouldn't do on CI), and on bug reports that contain test failures you can verify which version of dependencies were actually used in the run if a lock file is present.

**Comment by naderman** [comment]
By the way thanks for the PR! Just got distracted yesterday when I started the review so only left you with that one silly comment on the first file ;-)

File: src/Naderman/Composer/AWS/AwsClient.php [hunk 2]
```diff
@@ -125,39 +161,89 @@ public function determineBucketAndKey($url)
...
+             * If these are not set and we happen to have an IAM profile, it will still work.
              */
-            $key = getenv('AWS_ACCESS_KEY_ID');
-            $secret = getenv('AWS_SECRET_ACCESS_KEY');
-
-            if (false !== $key && false !== $secret) {
-                $s3config['key'] = $key;
-                $s3config['secret'] = $secret;
+            if (($composerAws = $config->get('amazon-aws'))) {
+                $s3config = array_merge($s3config, $composerAws);
```
Comment (naderman) [inline src/Naderman/Composer/AWS/AwsClient.php]: Do you think you can just stick in the "credentials" key here if needed, so that the config remains backward compatible?

File: tests/AwsClientTest.php [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+     * @param $bucket
+     * @param $key
+     */
+    public function testBucketAndKeyExtraction($url, $bucket, $key)
+    {
+        $client = new AwsClient($this->io, $this->config);
+        
+        $this->assertSame([$bucket, $key], $client->determineBucketAndKey($url));
+    }
+}
```
Comment (naderman) [inline tests/AwsClientTest.php]: Can you add newlines at the ends of file, this just always makes for silly diff changes later on ;-)

File: src/Naderman/Composer/AWS/AwsClient.php [hunk 4]
```diff
@@ -35,6 +36,16 @@ class AwsClient
     protected $io;
 
     /**
+     * @var S3Client
+     */
+    protected $client;
```
Comment (naderman) [inline src/Naderman/Composer/AWS/AwsClient.php]: why is this needed as state now? The s3factory function still returns the client, so I'd rather it wasn't state of this object?

Comment (naderman) [inline src/Naderman/Composer/AWS/AwsClient.php]: Seems like this is for tests only, can't you refactor the tests so this isn't needed?

**Comment by naderman** [comment]
Looks great apart from what I mentioned :-)

File: src/Naderman/Composer/AWS/AwsClient.php (hunk 4 above)
Comment (w3iBStime) [inline src/Naderman/Composer/AWS/AwsClient.php]: I'm going to see if I can address some of the issues in this PR.

It seems like @fdewinne keeps a reference to the client so that it doesn't have to get rebuilt for each download (e.g., packages.json + pack zip for every transitive dependency being loaded from S3).

**Comment by naderman** [comment]
Tagged this as https://github.com/naderman/composer-aws/releases/tag/1.0.0
